### PR TITLE
Allow registering custom arbs through GlobalArbResolver.

### LINKED ISSUE #3136: Allow registering custom default Arb for your own types in checkAll/forAll
Currently `checkAll` and `forAll` use `defaultForClass` and `targetDefaultForClass` to get a hardcoded list of global default Arb for those classes.

I have some more types (like `Vector3d` from JOML) that I would like to register default generators for, but it doesn't seem possible.

----

Maybe the definition of  `checkAll` could be amended like this:

```kt
suspend inline fun <reified A> checkAll(
   config: PropTestConfig,
   noinline property: suspend PropertyContext.(A) -> Unit
): PropertyContext = proptest(
   Arb.default<A>(config),
   config,
   property
)
```

### PR DESCRIPTION
**Pull request by sksamuel**
Closes #3136

### CODE REVIEW

File: kotest-property/src/commonMain/kotlin/io/kotest/property/resolution/ArbResolver.kt [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package io.kotest.property.resolution
+
+import io.kotest.property.Arb
+import kotlin.reflect.KType
+
+/**
+ * Looks up default [Arb] instances for types.
+ */
+interface ArbResolver {
```
Comment (Kantis) [inline kotest-property/src/commonMain/kotlin/io/kotest/property/resolution/ArbResolver.kt:9]:
```suggestion
internal interface ArbResolver {
```

File: kotest-property/src/commonMain/kotlin/io/kotest/property/resolution/ArbResolver.kt [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+   /**
+    * Returns a default [Arb] for the given [KType] or null if one is not available.
+    */
+   fun resolve(type: KType): Arb<*>?
+}
+
+object EmptyArbResolver : ArbResolver {
```
Comment (Kantis) [inline kotest-property/src/commonMain/kotlin/io/kotest/property/resolution/ArbResolver.kt:17]:
```suggestion
internal object EmptyArbResolver : ArbResolver {
```

File: kotest-property/src/commonMain/kotlin/io/kotest/property/resolution/resolve.kt [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import io.kotest.property.Arb
+import kotlin.reflect.KType
+import kotlin.reflect.typeOf
+
+/**
+ * Resolves a default [Arb] for reified type [A].
+ * See [resolve].
+ */
+@Suppress("UNCHECKED_CAST")
+inline fun <reified A> Arb.Companion.default(): Arb<A> = resolve(typeOf<A>()) as Arb<A>
```
Comment (Kantis) [inline kotest-property/src/commonMain/kotlin/io/kotest/property/resolution/resolve.kt:12]: I'm wondering if there's some way to use an `ArbResolver` as a context receiver here, to allow users to override Arb resolution.

```suggestion
context(ArbResolver)
inline fun <reified A> Arb.Companion.default(): Arb<A> = resolve(typeOf<A>()) as Arb<A> // should use ArbResolver#resolve
```

Ideally the method below (which could be considered a composite `ArbResolver` as well) would be added to the context by default.

**Review by Kantis — APPROVED** [review]

**Comment by sksamuel** [comment]
Oops auto merge merged this. I will apply your suggestions manually @Kantis

File: kotest-property/src/commonMain/kotlin/io/kotest/property/resolution/resolve.kt (hunk 3 above)
Comment (sksamuel) [reply kotest-property/src/commonMain/kotlin/io/kotest/property/resolution/resolve.kt:12]: I don't want to introduce context receivers to Kotest as that will move us to 1.8 as the minimum version (if context receivers are stable by then). We've been bitten before by experimental stuff so not yet.

File: kotest-property/src/commonMain/kotlin/io/kotest/property/resolution/ArbResolver.kt (hunk 2 above)
Comment (sksamuel) [reply kotest-property/src/commonMain/kotlin/io/kotest/property/resolution/ArbResolver.kt:17]: done

File: kotest-property/src/commonMain/kotlin/io/kotest/property/resolution/ArbResolver.kt (hunk 1 above)
Comment (sksamuel) [reply kotest-property/src/commonMain/kotlin/io/kotest/property/resolution/ArbResolver.kt:9]: done

**Pull request merged**